File: compiled/cpp/nanocode.cpp
// nanocode - minimal claude code alternative (C++)
// g++ -std=c++17 -o nanocode nanocode.cpp -lcurl && ./nanocode
#include <iostream>
#include <fstream>
#include <sstream>
#include <string>
#include <vector>
#include <curl/curl.h>
#include <cstdlib>
#include <regex>

#define R "\033[0m"
#define B "\033[1m"
#define D "\033[2m"
#define C "\033[36m"
#define G "\033[32m"
#define BL "\033[34m"

std::string KEY, MODEL, API;
std::string msgs = "[]";

size_t write_cb(void* p, size_t s, size_t n, std::string* out) {
    out->append((char*)p, s * n);
    return s * n;
}

std::string run_tool(const std::string& name, const std::string& input) {
    std::regex path_re(R"("path"\s*:\s*"([^"]*)")");
    std::regex cmd_re(R"("cmd"\s*:\s*"([^"]*)")");
    std::regex pat_re(R"("pat"\s*:\s*"([^"]*)")");
    std::regex old_re(R"("old"\s*:\s*"([^"]*)")");
    std::regex new_re(R"("new"\s*:\s*"([^"]*)")");
    std::regex content_re(R"("content"\s*:\s*"([^"]*)")");
    std::regex off_re(R"("offset"\s*:\s*(\d+))");
    std::regex lim_re(R"("limit"\s*:\s*(\d+))");
    std::smatch m;
    
    if (name == "read" && std::regex_search(input, m, path_re)) {
        std::ifstream f(m[1]);
        std::string result, line;
        int off = std::regex_search(input, m, off_re) ? std::stoi(m[1]) : 0;
        int lim = std::regex_search(input, m, lim_re) ? std::stoi(m[1]) : 0;
        int line_no = 0, emitted = 0;
        while (std::getline(f, line) && (lim == 0 || emitted < lim)) {
            if (line_no++ < off) continue;
            result += std::to_string(line_no) + "| " + line + "\n";
            emitted++;
        }
        return result;
    }
    if (name == "write" && std::regex_search(input, m, path_re) && std::regex_search(input, m, content_re)) {
        std::string path = m[1];
        std::regex_search(input, m, content_re);
        std::string content = m[1];
        std::ofstream f(path); f << content; f.close(); return "ok";
    }
    if (name == "edit" && std::regex_search(input, m, path_re) && std::regex_search(input, m, old_re) && std::regex_search(input, m, new_re)) {
        std::string path = m[1], oldv = m[1], newv;
        std::regex_search(input, m, old_re); oldv = m[1];
        std::regex_search(input, m, new_re); newv = m[1];
        bool all = input.find("\"all\":true") != std::string::npos;
        std::ifstream f(path); if (!f) return "error: file not found";
        std::stringstream buf; buf << f.rdbuf(); std::string txt = buf.str(); f.close();
        size_t first = txt.find(oldv);
        if (first == std::string::npos) return "error: old_string not found";
        if (!all && txt.find(oldv, first + oldv.size()) != std::string::npos) return "error: old_string appears multiple times, use all=true";
        if (all) {
            size_t pos = 0;
            while ((pos = txt.find(oldv, pos)) != std::string::npos) { txt.replace(pos, oldv.size(), newv); pos += newv.size(); }
        } else {
            txt.replace(first, oldv.size(), newv);
        }
        std::ofstream o(path); o << txt; o.close();
        return "ok";
    }
    if (name == "bash" && std::regex_search(input, m, cmd_re)) {
        char buf[256];
        std::string result;
        FILE* fp = popen(m[1].str().c_str(), "r");
        while (fp && fgets(buf, sizeof(buf), fp)) result += buf;
        if (fp) pclose(fp);
        return result;
    }
    if (name == "glob" && std::regex_search(input, m, pat_re)) {
        char buf[256];
        std::string result;
        std::string cmd = "find . -name '" + m[1].str() + "' | head -50";
        FILE* fp = popen(cmd.c_str(), "r");
        while (fp && fgets(buf, sizeof(buf), fp)) result += buf;
        if (fp) pclose(fp);
        return result;
    }
    if (name == "grep" && std::regex_search(input, m, pat_re)) {
        char buf[256];
        std::string result;
        std::string cmd = "grep -R \"" + m[1].str() + "\" . -n 2>/dev/null | head -50";
        FILE* fp = popen(cmd.c_str(), "r");
        while (fp && fgets(buf, sizeof(buf), fp)) result += buf;
        if (fp) pclose(fp);
        return result.empty() ? "none" : result;
    }
    return "ok";
}

std::string api_call() {
    std::string body = R"({"model":")" + MODEL + R"(","max_tokens":4096,"system":"Concise assistant","messages":)" + msgs + R"(,)"
        R"("tools":[)"
        R"({"name":"read","description":"Read","input_schema":{"type":"object","properties":{"path":{"type":"string"},"offset":{"type":"integer"},"limit":{"type":"integer"}},"required":["path"]}},)"
        R"({"name":"write","description":"Write","input_schema":{"type":"object","properties":{"path":{"type":"string"},"content":{"type":"string"}},"required":["path","content"]}},)"
        R"({"name":"edit","description":"Replace","input_schema":{"type":"object","properties":{"path":{"type":"string"},"old":{"type":"string"},"new":{"type":"string"},"all":{"type":"boolean"}},"required":["path","old","new"]}},)"
        R"({"name":"bash","description":"Run","input_schema":{"type":"object","properties":{"cmd":{"type":"string"}},"required":["cmd"]}},)"
        R"({"name":"glob","description":"Find","input_schema":{"type":"object","properties":{"pat":{"type":"string"}},"required":["pat"]}},)"
        R"({"name":"grep","description":"Search","input_schema":{"type":"object","properties":{"pat":{"type":"string"}},"required":["pat"]}}]})";
    
    CURL* curl = curl_easy_init();
    std::string resp;
    struct curl_slist* hdrs = nullptr;
    hdrs = curl_slist_append(hdrs, "Content-Type: application/json");
    hdrs = curl_slist_append(hdrs, "anthropic-version: 2023-06-01");
    if (const char* ok = std::getenv("OPENROUTER_API_KEY")) {
        hdrs = curl_slist_append(hdrs, ("Authorization: Bearer " + std::string(ok)).c_str());
    } else {
        hdrs = curl_slist_append(hdrs, ("x-api-key: " + KEY).c_str());
    }
    
    curl_easy_setopt(curl, CURLOPT_URL, API.c_str());
    curl_easy_setopt(curl, CURLOPT_HTTPHEADER, hdrs);
    curl_easy_setopt(curl, CURLOPT_POSTFIELDS, body.c_str());
    curl_easy_setopt(curl, CURLOPT_WRITEFUNCTION, write_cb);
    curl_easy_setopt(curl, CURLOPT_WRITEDATA, &resp);
    curl_easy_perform(curl);
    curl_easy_cleanup(curl);
    curl_slist_free_all(hdrs);
    return resp;
}

int main() {
    const char* or_key = std::getenv("OPENROUTER_API_KEY");
    const char* an_key = std::getenv("ANTHROPIC_API_KEY");
    KEY = or_key ? or_key : (an_key ? an_key : "");
    API = or_key ? "https://openrouter.ai/api/v1/messages" : "https://api.anthropic.com/v1/messages";
    MODEL = std::getenv("MODEL") ?: (or_key ? "anthropic/claude-opus-4-5" : "claude-opus-4-5");
    std::cout << B "nanocode" R " | " D << MODEL << R "\n\n";
    
    std::string input;
    while (true) {
        std::cout << B BL "❯" R " " << std::flush;
        if (!std::getline(std::cin, input)) break;
        while (!input.empty() && (input.back() == ' ' || input.back() == '\n')) input.pop_back();
        if (input.empty()) continue;
        if (input == "/q") break;
        if (input == "/c") { msgs = "[]"; std::cout << G "⏺ Cleared" R "\n"; continue; }
        
        // Escape input for JSON
        std::string escaped;
        for (char c : input) {
            if (c == '"') escaped += "\\\"";
            else if (c == '\\') escaped += "\\\\";
            else if (c == '\n') escaped += "\\n";
            else escaped += c;
        }
        msgs = msgs.substr(0, msgs.length()-1) + R"({"role":"user","content":")" + escaped + R"("}])";
        
        std::string resp = api_call();
        
        // Extract text
        std::regex text_re(R"("type"\s*:\s*"text"[^}]*"text"\s*:\s*"([^"]*)")");
        std::smatch m;
        if (std::regex_search(resp, m, text_re))
            std::cout << "\n" C "⏺" R " " << m[1] << "\n";
        
        // Extract tool use
        std::regex tool_re(R"("type"\s*:\s*"tool_use"[^}]*"name"\s*:\s*"(\w+)"[^}]*"input"\s*:\s*\{([^}]*)\})");
        if (std::regex_search(resp, m, tool_re)) {
            std::string name = m[1], inp = "{" + m[2].str() + "}";
            std::cout << "\n" G "⏺ " << name << R "\n";
            std::string result = run_tool(name, inp);
            std::cout << "  " D "⎿ " << result.substr(0, result.find('\n')) << R "\n";
        }
        std::cout << "\n";
    }
    return 0;
}

File: compiled/zig/nanocode.zig
// nanocode - minimal claude code alternative (Zig)
// zig build-exe nanocode.zig && ./nanocode
const std = @import("std");

const R = "\x1b[0m";
const B = "\x1b[1m";
const D = "\x1b[2m";
const C = "\x1b[36m";
const G = "\x1b[32m";
const BL = "\x1b[34m";
const RD = "\x1b[31m";

const Allocator = std.mem.Allocator;

pub fn main() !void {
    var gpa = std.heap.GeneralPurposeAllocator(.{}){};
    const alloc = gpa.allocator();

    const or_key = std.posix.getenv("OPENROUTER_API_KEY");
    const an_key = std.posix.getenv("ANTHROPIC_API_KEY");
    const key = or_key orelse an_key orelse @panic("ANTHROPIC_API_KEY or OPENROUTER_API_KEY required");
    const api = if (or_key != null) "https://openrouter.ai/api/v1/messages" else "https://api.anthropic.com/v1/messages";
    const default_model = if (or_key != null) "anthropic/claude-opus-4" else "claude-sonnet-4-20250514";
    const model = std.posix.getenv("MODEL") orelse default_model;
    const auth_header = if (or_key != null) try std.fmt.allocPrint(alloc, "Authorization: Bearer {s}", .{key}) else try std.fmt.allocPrint(alloc, "x-api-key: {s}", .{key});

    const stdout = std.io.getStdOut().writer();
    const stdin = std.io.getStdIn().reader();

    var cwd_buf: [std.fs.max_path_bytes]u8 = undefined;
    const cwd = std.fs.cwd().realpath(".", &cwd_buf) catch ".";

    try stdout.print("{s}nanocode{s} | {s}{s} | {s}{s}\n\n", .{ B, R, D, model, cwd, R });

    var messages = std.ArrayList([]const u8).init(alloc);
    var buf: [8192]u8 = undefined;

    while (true) {
        try stdout.print("{s}{s}❯{s} ", .{ B, BL, R });

        const line = stdin.readUntilDelimiterOrEof(&buf, '\n') catch break;
        if (line == null) break;

        const input = std.mem.trim(u8, line.?, " \t\r\n");
        if (input.len == 0) continue;
        if (std.mem.eql(u8, input, "/q") or std.mem.eql(u8, input, "exit")) break;
        if (std.mem.eql(u8, input, "/c")) {
            messages.clearRetainingCapacity();
            try stdout.print("{s}⏺ Cleared{s}\n", .{ G, R });
            continue;
        }

        try messages.append(try std.fmt.allocPrint(alloc, "{{\"role\":\"user\",\"content\":{s}}}", .{try jsonStr(alloc, input)}));

        while (true) {
            const msgs_json = try std.mem.join(alloc, ",", messages.items);
            const body = try std.fmt.allocPrint(alloc,
                \\{{"model":"{s}","max_tokens":8192,"system":"Concise coding assistant. cwd: {s}","messages":[{s}],"tools":[{s},{s},{s},{s},{s},{s}]}}
            , .{ model, cwd, msgs_json, tool_read, tool_write, tool_edit, tool_glob, tool_grep, tool_bash });

            const result = try std.process.Child.run(.{
                .allocator = alloc,
                .argv = &.{ "curl", "-s", api, "-H", "Content-Type: application/json", "-H", "anthropic-version: 2023-06-01", "-H", auth_header, "-d", body },
            });

            const resp = result.stdout;
            var has_tool_use = false;
            var tool_results = std.ArrayList([]const u8).init(alloc);
            var content_parts = std.ArrayList([]const u8).init(alloc);

            // Parse content blocks
            var pos: usize = 0;
            while (std.mem.indexOfPos(u8, resp, pos, "\"type\":\"")) |type_start| {
                const t_start = type_start + 8;
                if (std.mem.indexOfPos(u8, resp, t_start, "\"")) |t_end| {
                    const block_type = resp[t_start..t_end];

                    if (std.mem.eql(u8, block_type, "text")) {
                        if (extractField(resp, t_end, "text")) |text| {
                            try stdout.print("\n{s}⏺{s} {s}", .{ C, R, unescape(alloc, text) catch text });
                            try content_parts.append(try std.fmt.allocPrint(alloc, "{{\"type\":\"text\",\"text\":{s}}}", .{try jsonStr(alloc, text)}));
                        }
                    } else if (std.mem.eql(u8, block_type, "tool_use")) {
                        has_tool_use = true;
                        const id = extractField(resp, t_end, "id") orelse "unknown";
                        const name = extractField(resp, t_end, "name") orelse "unknown";
                        const input_json = extractInput(resp, t_end) orelse "{}";

                        try stdout.print("\n{s}⏺ {s}{s}", .{ G, name, R });
                        const tool_result = runTool(alloc, name, input_json) catch |e| try std.fmt.allocPrint(alloc, "error: {}", .{e});
                        const first_line = std.mem.sliceTo(tool_result, '\n');
                        const preview = if (first_line.len > 60) first_line[0..60] else first_line;
                        try stdout.print("\n  {s}⎿ {s}{s}", .{ D, preview, R });

                        try content_parts.append(try std.fmt.allocPrint(alloc, "{{\"type\":\"tool_use\",\"id\":\"{s}\",\"name\":\"{s}\",\"input\":{s}}}", .{ id, name, input_json }));
                        try tool_results.append(try std.fmt.allocPrint(alloc, "{{\"type\":\"tool_result\",\"tool_use_id\":\"{s}\",\"content\":{s}}}", .{ id, try jsonStr(alloc, tool_result) }));
                    }
                    pos = t_end + 1;
                } else break;
            }

            if (content_parts.items.len > 0) {
                const content_json = try std.mem.join(alloc, ",", content_parts.items);
                try messages.append(try std.fmt.allocPrint(alloc, "{{\"role\":\"assistant\",\"content\":[{s}]}}", .{content_json}));
            }

            if (has_tool_use and tool_results.items.len > 0) {
                const results_json = try std.mem.join(alloc, ",", tool_results.items);
                try messages.append(try std.fmt.allocPrint(alloc, "{{\"role\":\"user\",\"content\":[{s}]}}", .{results_json}));
            } else {
                break;
            }
        }
        try stdout.print("\n\n", .{});
    }
}

const tool_read =
    \\{"name":"read","description":"Read file","input_schema":{"type":"object","properties":{"path":{"type":"string"},"offset":{"type":"integer"},"limit":{"type":"integer"}},"required":["path"]}}
;
const tool_write =
    \\{"name":"write","description":"Write file","input_schema":{"type":"object","properties":{"path":{"type":"string"},"content":{"type":"string"}},"required":["path","content"]}}
;
const tool_edit =
    \\{"name":"edit","description":"Edit file","input_schema":{"type":"object","properties":{"path":{"type":"string"},"old":{"type":"string"},"new":{"type":"string"},"all":{"type":"boolean"}},"required":["path","old","new"]}}
;
const tool_glob =
    \\{"name":"glob","description":"Find files","input_schema":{"type":"object","properties":{"pat":{"type":"string"},"path":{"type":"string"}},"required":["pat"]}}
;
const tool_grep =
    \\{"name":"grep","description":"Search files","input_schema":{"type":"object","properties":{"pat":{"type":"string"},"path":{"type":"string"}},"required":["pat"]}}
;
const tool_bash =
    \\{"name":"bash","description":"Run command","input_schema":{"type":"object","properties":{"cmd":{"type":"string"}},"required":["cmd"]}}
;

fn extractField(resp: []const u8, start: usize, field: []const u8) ?[]const u8 {
    const search = std.fmt.allocPrint(std.heap.page_allocator, "\"{s}\":\"", .{field}) catch return null;
    if (std.mem.indexOfPos(u8, resp, start, search)) |f_start| {
        const v_start = f_start + search.len;
        var end = v_start;
        while (end < resp.len) : (end += 1) {
            if (resp[end] == '"' and (end == v_start or resp[end - 1] != '\\')) break;
        }
        return resp[v_start..end];
    }
    return null;
}

fn extractInput(resp: []const u8, start: usize) ?[]const u8 {
    if (std.mem.indexOfPos(u8, resp, start, "\"input\":")) |i_start| {
        const v_start = i_start + 8;
        if (v_start < resp.len and resp[v_start] == '{') {
            var depth: i32 = 1;
            var end = v_start + 1;
            while (end < resp.len and depth > 0) : (end += 1) {
                if (resp[end] == '{') depth += 1 else if (resp[end] == '}') depth -= 1;
            }
            return resp[v_start..end];
        }
    }
    return null;
}

fn runTool(alloc: Allocator, name: []const u8, input_json: []const u8) ![]const u8 {
    if (std.mem.eql(u8, name, "bash")) {
        const cmd = extractJsonField(input_json, "cmd") orelse return "error: no cmd";
        const result = try std.process.Child.run(.{ .allocator = alloc, .argv = &.{ "sh", "-c", cmd } });
        return if (result.stdout.len > 0) result.stdout else if (result.stderr.len > 0) result.stderr else "(empty)";
    } else if (std.mem.eql(u8, name, "read")) {
        const path = extractJsonField(input_json, "path") orelse return "error: no path";
        const file = std.fs.cwd().openFile(path, .{}) catch return "error: file not found";
        defer file.close();
        return file.readToEndAlloc(alloc, 1024 * 1024) catch "error: read failed";
    } else if (std.mem.eql(u8, name, "write")) {
        const path = extractJsonField(input_json, "path") orelse return "error: no path";
        const content = extractJsonField(input_json, "content") orelse return "error: no content";
        const file = std.fs.cwd().createFile(path, .{}) catch return "error: create failed";
        defer file.close();
        file.writeAll(try unescape(alloc, content)) catch return "error: write failed";
        return "ok";
    } else if (std.mem.eql(u8, name, "edit")) {
        const path = extractJsonField(input_json, "path") orelse return "error: no path";
        const old = try unescape(alloc, extractJsonField(input_json, "old") orelse return "error: no old");
        const new = try unescape(alloc, extractJsonField(input_json, "new") orelse return "error: no new");
        const file = std.fs.cwd().openFile(path, .{}) catch return "error: file not found";
        const content = file.readToEndAlloc(alloc, 1024 * 1024) catch return "error: read failed";
        file.close();
        if (std.mem.indexOf(u8, content, old) == null) return "error: not found";
        const replaced = try std.mem.replaceOwned(u8, alloc, content, old, new);
        const out = std.fs.cwd().createFile(path, .{}) catch return "error: create failed";
        defer out.close();
        out.writeAll(replaced) catch return "error: write failed";
        return "ok";
    } else if (std.mem.eql(u8, name, "glob")) {
        const pat = extractJsonField(input_json, "pat") orelse return "error: no pat";
        const result = try std.process.Child.run(.{ .allocator = alloc, .argv = &.{ "sh", "-c", try std.fmt.allocPrint(alloc, "find . -name '{s}' 2>/dev/null | head -50", .{pat}) } });
        return if (result.stdout.len > 0) result.stdout else "none";
    } else if (std.mem.eql(u8, name, "grep")) {
        const pat = extractJsonField(input_json, "pat") orelse return "error: no pat";
        const result = try std.process.Child.run(.{ .allocator = alloc, .argv = &.{ "sh", "-c", try std.fmt.allocPrint(alloc, "grep -rn '{s}' . 2>/dev/null | head -50", .{pat}) } });
        return if (result.stdout.len > 0) result.stdout else "none";
    }
    return "error: unknown tool";
}

fn extractJsonField(json: []const u8, field: []const u8) ?[]const u8 {
    const search = std.fmt.allocPrint(std.heap.page_allocator, "\"{s}\":\"", .{field}) catch return null;
    if (std.mem.indexOf(u8, json, search)) |start| {
        const v_start = start + search.len;
        var end = v_start;
        while (end < json.len) : (end += 1) {
            if (json[end] == '"' and (end == v_start or json[end - 1] != '\\')) break;
        }
        return json[v_start..end];
    }
    return null;
}

fn jsonStr(alloc: Allocator, s: []const u8) ![]const u8 {
    var result = std.ArrayList(u8).init(alloc);
    try result.append('"');
    for (s) |c| {
        switch (c) {
            '"' => try result.appendSlice("\\\""),
            '\\' => try result.appendSlice("\\\\"),
            '\n' => try result.appendSlice("\\n"),
            '\r' => try result.appendSlice("\\r"),
            '\t' => try result.appendSlice("\\t"),
            else => try result.append(c),
        }
    }
    try result.append('"');
    return result.items;
}

fn unescape(alloc: Allocator, s: []const u8) ![]const u8 {
    var result = std.ArrayList(u8).init(alloc);
    var i: usize = 0;
    while (i < s.len) : (i += 1) {
        if (s[i] == '\\' and i + 1 < s.len) {
            switch (s[i + 1]) {
                'n' => try result.append('\n'),
                'r' => try result.append('\r'),
                't' => try result.append('\t'),
                '"' => try result.append('"'),
                '\\' => try result.append('\\'),
                else => try result.append(s[i + 1]),
            }
            i += 1;
        } else try result.append(s[i]);
    }
    return result.items;
}
